analyses/chromothripsis/02-run-shatterseek-and-classify-confidence.R

```diff
@@ -0,0 +1,293 @@
+# There can be gaps of any length in between the segments - they just need to be consecutive (not adjacent) 
+# on the same chromosome and have the same copy number value.
+
+# I couldn't find a solution to do something like this using established packages, but I'm happy to hear suggestions 
+# if anyone can think of a way to simplify this.
+
+mergeCNsegments <- function(cnv_df) {
+
+  # Order rows by chromosome and start position 
+  cnv_df <- cnv_df[with(cnv_df, order(chrom, loc.start)),]
+  
+  # Annotate each row with TRUE if its chromosome and CN value matches the next row, or FALSE if it doesn't
+  checkCNV_logical <- logical()
+  num_rows <- dim(cnv_df)[1]-1 # Subtract 1 because the last row will not have a comparison
+  for (n in 1:num_rows) {
+    checkCNV_logical[n] <-
+      (cnv_df[n, "chrom"] == cnv_df[n+1, "chrom"]) & 
+      (cnv_df[n, "copy.num"] == cnv_df[n+1, "copy.num"])
+  }
+  cnv_df$checkCNV_logical <- c(checkCNV_logical, NA) # Append an NA for the last row
+  
+  # Now identify runs of TRUE values, and extract their start and end positions within the dataframe
+  # to get the positions for consecutive rows with matching copy number
+  # (Adapted from https://masterr.org/r/how-to-find-consecutive-repeats-in-r/)
+  
+  # Identify consecutive runs of same value (TRUE or FALSE)
+  runs <- rle(cnv_df$checkCNV_logical)
+  
+  # Create index for TRUE runs 
+  runs_TRUE <- which(runs$values == TRUE)
+  
+  # Calculate cumulative sum for run lengths
+  runs.lengths.cumsum <- cumsum(runs$lengths)
+  
+  # Find end positions for TRUE runs; add 1 to get the last row with matching CN
+  # (because TRUE is defined based on row n and row n+1)
+  ends <- runs.lengths.cumsum[runs_TRUE] + 1
+  
+  # Find start positions for TRUE runs
+  newindex <- ifelse(runs_TRUE>1, runs_TRUE-1, 0)
+  starts <- runs.lengths.cumsum[newindex] + 1
+  if (0 %in% newindex) starts <- c(1,starts)
+  
+  # Loop through start and end positions for consecutive rows of matching CN; 
+  # create new merged rows; define a list of unmerged rows to remove
+  merged_rows <- data.frame()
+  rows_to_remove <- vector()
+  for (i in 1:length(starts)){
+    merged_chrom <- cnv_df[starts[i], "chrom"]
+    merged_start <- cnv_df[starts[i], "loc.start"]
+    merged_end <- cnv_df[ends[i], "loc.end"]
+    merged_cn <- cnv_df[starts[i], "copy.num"]
+    merged_row <- data.frame(chrom=merged_chrom, loc.start=merged_start, loc.end=merged_end, copy.num=merged_cn)
+    merged_rows <- rbind(merged_rows, merged_row)
+    rows_to_remove <- c(rows_to_remove, starts[i]:ends[i])
+  }
+  
+  # Remove unmerged rows from CNV df
+  cnv_df_merged <- cnv_df[-rows_to_remove,]
+  
+  # Add merged rows to CNV df (first subset to the necessary columns)
+  cnv_df_merged <- cnv_df_merged[,c(2:4,7)]
+  cnv_df_merged <- rbind(cnv_df_merged, merged_rows)
+  
+  # Reorder rows by chromosome and start position 
+  cnv_df_merged <- cnv_df_merged[with(cnv_df_merged, order(chrom, loc.start)),]
```

Comment:
Hmm, it's not explicitly noted in the documentation - but from my understanding of how the functions run, I think the rows do need to be ordered by loc.start within each chromosome. I also ordered it so that I could compare cnv_df_merged to cnv_df more easily while testing.